File: SGFastfish/SGItemsView.m
//
//  SGItemsView.m
//  SGFastfishExample
//
//  Created by kingsic on 2020/10/15.
//  Copyright © 2020 kingsic. All rights reserved.
//

#import "SGItemsView.h"

@interface SGItemCell : UICollectionViewCell
@property (nonatomic, strong) UIImageView *imgView;
@property (nonatomic, strong) UILabel *titleLab;
@end
@implementation SGItemCell

- (instancetype)initWithFrame:(CGRect)frame {
    if (self = [super initWithFrame:frame]) {
        self.backgroundColor = [UIColor clearColor];
        [self addSubview:self.imgView];
        [self addSubview:self.titleLab];
    }
    return self;
}
- (void)layoutSubviews {
    [super layoutSubviews];
    
    CGFloat iv_y = 5;
    CGFloat iv_height = self.frame.size.height / 3 * 2 - iv_y;
    CGFloat iv_width = iv_height;
    CGFloat iv_x = 0.5 * (self.frame.size.width - iv_width);
    _imgView.frame = CGRectMake(iv_x, iv_y, iv_width, iv_height);
    
    CGFloat tl_x = 0;
    CGFloat tl_y = self.frame.size.height / 3 * 2;
    CGFloat tl_width = self.frame.size.width;
    CGFloat tl_height = self.frame.size.height / 3;
    _titleLab.frame = CGRectMake(tl_x, tl_y, tl_width, tl_height);
}

- (UIImageView *)imgView {
    if (!_imgView) {
        _imgView = [[UIImageView alloc] init];
    }
    return _imgView;
}
- (UILabel *)titleLab {
    if (!_titleLab) {
        _titleLab = [[UILabel alloc] init];
        _titleLab.textAlignment = NSTextAlignmentCenter;
    }
    return _titleLab;
}
@end


@interface SGItemsView () <UICollectionViewDelegate, UICollectionViewDataSource>
@property (nonatomic, strong) UICollectionViewFlowLayout *CVFlowLayout;
@property (nonatomic, strong) UICollectionView *collectionView;
@property (nonatomic, strong) NSMutableDictionary *itemTitleColorMDict;
@end

@implementation SGItemsView

- (instancetype)initWithFrame:(CGRect)frame {
    if (self = [super initWithFrame:frame]) {
        [self configure];
        [self addSubview:self.collectionView];
    }
    return self;
}
- (void)configure {
    _titleFont = [UIFont systemFontOfSize:12];
}

- (void)layoutSubviews {
    [super layoutSubviews];
    
    _CVFlowLayout.itemSize = CGSizeMake(_itemSize.width, _itemSize.height);
    
    CGFloat cv_y = 0;
    CGFloat cv_width = self.frame.size.width;
    CGFloat cv_height = self.frame.size.height;
    _collectionView.frame = CGRectMake(0, cv_y, cv_width, cv_height);
}

- (UICollectionView *)collectionView {
    if (!_collectionView) {
        _CVFlowLayout = [[UICollectionViewFlowLayout alloc] init];
        _CVFlowLayout.minimumInteritemSpacing = 0;
        _CVFlowLayout.minimumLineSpacing = 0;
        
        _collectionView = [[UICollectionView alloc] initWithFrame:CGRectZero collectionViewLayout:_CVFlowLayout];
        _collectionView.delegate = self;
        _collectionView.dataSource = self;
        [_collectionView registerClass:[SGItemCell class] forCellWithReuseIdentifier:@"cellID"];
        _collectionView.backgroundColor = self.backgroundColor;
        _collectionView.showsHorizontalScrollIndicator = _showsHorizontalScrollIndicator;
    }
    return _collectionView;
}
#pragma mark - - UICollectionViewDataSource
- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {
    return _titles.count;
}
- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath {
    SGItemCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:@"cellID" forIndexPath:indexPath];
    if (self.configureImgViewBlock) {
        self.configureImgViewBlock(cell.imgView, indexPath.row);
    }
    
    cell.titleLab.text = _titles[indexPath.item];
    if (_titleFont) {
        cell.titleLab.font = _titleFont;
    }
    if (_titleColor) {
        cell.titleLab.textColor = _titleColor;
    }

    if (_itemTitleColorMDict) {
        [self.itemTitleColorMDict enumerateKeysAndObjectsUsingBlock:^(id  _Nonnull key, id  _Nonnull obj, BOOL * _Nonnull stop) {
            NSInteger index = [key integerValue];
            if (index == indexPath.row) {
                cell.titleLab.textColor = obj;
            }
        }];
    }
    return cell;
}
- (void)collectionView:(UICollectionView *)collectionView didSelectItemAtIndexPath:(NSIndexPath *)indexPath {
    if (self.itemClickBlock) {
        self.itemClickBlock(indexPath.row);
    }
}

#pragma mark - - set
- (void)setTitles:(NSArray *)titles {
    _titles = titles;
    [_collectionView reloadData];
}
- (void)setTitleFont:(UIFont *)titleFont {
    _titleFont = titleFont;
}
- (void)setTitleColor:(UIColor *)titleColor {
    _titleColor = titleColor;
}

- (void)setItemSize:(CGSize)itemSize {
    _itemSize = itemSize;
}
- (void)setContentInset:(UIEdgeInsets)contentInset {
    _collectionView.contentInset = UIEdgeInsetsMake(contentInset.top, contentInset.left, contentInset.bottom, contentInset.right);
}
- (void)setPagingEnabled:(BOOL)pagingEnabled {
    _collectionView.pagingEnabled = pagingEnabled;
}
- (void)setScrollDirectionHorizontal:(BOOL)scrollDirectionHorizontal {
    if (scrollDirectionHorizontal == YES) {
        _CVFlowLayout.scrollDirection = UICollectionViewScrollDirectionHorizontal;
    }
}
- (void)setShowsHorizontalScrollIndicator:(BOOL)showsHorizontalScrollIndicator {
    _collectionView.showsHorizontalScrollIndicator = showsHorizontalScrollIndicator;
}
- (void)setShowsVerticalScrollIndicator:(BOOL)showsVerticalScrollIndicator {
    _collectionView.showsVerticalScrollIndicator = showsVerticalScrollIndicator;
}

- (void)setItemTitleColor:(UIColor *)titleColor forIndex:(NSInteger)index {
    [self.itemTitleColorMDict setObject:titleColor forKey:@(index)];
}
- (NSMutableDictionary *)itemTitleColorMDict {
    if (!_itemTitleColorMDict) {
        _itemTitleColorMDict = [NSMutableDictionary dictionary];
    }
    return _itemTitleColorMDict;
}

@end

File: Flower/Sources/UIKit/SGItemsView.swift
//
//  SGItemsView.swift
//  SGItemsView
//
//  Created by kingsic on 2020/10/23.
//  Copyright © 2020 kingsic. All rights reserved.
//

import UIKit

public class SGItemsView: UIView {
    /// Item titles
    public var titles: Array<Any>?
    
    /// Title font，default system font 12 plain
    public var titleFont: UIFont = .systemFont(ofSize: 12)
    
    /// Title color，default black
    public var titleColor: UIColor = .black
    
    public typealias ConfigureImgViewBlock = (_ imageView: UIImageView, _ index: Int) -> Void
    private var tempConfigureImgViewBlock: ConfigureImgViewBlock?
    /// Configure imageView callback method, leave the imageView to external processing, such as network or local loading image
    public func configureImgViewBlock(imgViewBlock: @escaping ConfigureImgViewBlock) {
        tempConfigureImgViewBlock = imgViewBlock
    }
    
    /// Item size. consider the top and bottom values of the contentinset property when setting the height
    public var itemSize: CGSize?
    
    /// Item content Inset of top, defalut is 5.0
    ///
    /// The maximum value cannot exceed 1/6 of itemSize
    public var contentInsetTop: CGFloat = 5.0
    
    /// Item content Inset of bottom, defalut is 5.0
    ///
    /// The maximum value cannot exceed 1/6 of itemSize
    public var contentInsetBottom: CGFloat = 5.0

    /// Item space between image and title, defalut is 5.0
    ///
    /// The maximum value cannot exceed 1/6 of itemSize
    public var spacing: CGFloat = 5.0
    
    /// Default UIEdgeInsetsZero. add additional scroll area around content
    public var contentInset: UIEdgeInsets {
        set {
            collectionView.contentInset = UIEdgeInsets.init(top: newValue.top, left: newValue.left, bottom: newValue.bottom, right: newValue.right)
        }
        get {
            return .zero
        }
    }
    
    /// Default false. if true, stop on multiples of view bounds
    public var pagingEnabled: Bool {
        set {
            if newValue {
                collectionView.isPagingEnabled = newValue
            }
        }
        get {
            return false
        }
    }
    
    /// Default false. if true, scroll direction Horizontal
    public var scrollDirectionHorizontal: Bool? {
        willSet {
            if newValue == true {
                CVFLayout.scrollDirection = .horizontal
            }
        }
    }
    
    /// Default false. if true, show indicator while we are tracking. fades out after tracking
    public var showsHorizontalScrollIndicator: Bool {
        set {
            if newValue == true {
                collectionView.showsHorizontalScrollIndicator = true
            }
        }
        get {
            return false
        }
    }
    
    public typealias ItemClickBlock = (_ index: Int) -> Void
    private var tempItemClickBlock: ItemClickBlock?
    /// Item click callback method
    public func itemClickBlock(itemClick: ItemClickBlock?) {
        tempItemClickBlock = itemClick
    }
    
    var itemTitleColorDict = [String: UIColor]()
    /// Set the titleColor of item according to the subscript
    public func setItemTitle(color: UIColor, index: Int) {
        itemTitleColorDict["\(index)"] = color
    }
    

    override init(frame: CGRect) {
        super.init(frame: frame)
        addSubview(collectionView)
    }
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    public override func layoutSubviews() {
        super.layoutSubviews()
        if let tempItemSize = itemSize {
            CVFLayout.itemSize = CGSize(width: tempItemSize.width, height: tempItemSize.height)
        }
        
        let cv_x: CGFloat = 0.0
        let cv_y: CGFloat = 0.0
        let cv_w = self.frame.size.width
        let cv_h = self.frame.size.height
        collectionView.frame = CGRect.init(x: cv_x, y: cv_y, width: cv_w, height: cv_h)
    }
    
    private lazy var CVFLayout: UICollectionViewFlowLayout = {
        let tempCVFLayout = UICollectionViewFlowLayout()
        tempCVFLayout.minimumLineSpacing = 0
        tempCVFLayout.minimumInteritemSpacing = 0
        return tempCVFLayout
    }()
    private lazy var collectionView: UICollectionView = {
        let tempCollectionView = UICollectionView(frame: .zero, collectionViewLayout: CVFLayout)
        tempCollectionView.delegate = self
        tempCollectionView.dataSource = self
        tempCollectionView.register(SGItemCell.self, forCellWithReuseIdentifier: "cellID")
        tempCollectionView.backgroundColor = self.backgroundColor
        tempCollectionView.showsHorizontalScrollIndicator = false
        return tempCollectionView
    }()
    
    
}

extension SGItemsView: UICollectionViewDelegate, UICollectionViewDataSource {
    public func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return titles?.count ?? 0
    }
    
    public func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "cellID", for: indexPath) as! SGItemCell
        if let configureImgViewBlock = tempConfigureImgViewBlock {
            configureImgViewBlock(cell.imgView, indexPath.item)
        }
        cell.contentInsetOfTop = contentInsetTop
        cell.contentInsetOfBottom = contentInsetBottom
        cell.spacing = spacing
        cell.titleLab.text = (titles?[indexPath.item] as! String)
        cell.titleLab.font = titleFont
        cell.titleLab.textColor = titleColor
        if !itemTitleColorDict.isEmpty {
            for (key, value) in itemTitleColorDict {
                if Int(key) == indexPath.item {
                    cell.titleLab.textColor = value
                }
            }
        }
        return cell
    }
    
    public func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
        if let itemClickBlock = tempItemClickBlock {
            itemClickBlock(indexPath.item)
        }
    }
}


private class SGItemCell: UICollectionViewCell {
    public var contentInsetOfTop: CGFloat = 0.0
    public var contentInsetOfBottom: CGFloat = 0.0
    public var spacing: CGFloat = 0.0
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        addSubview(imgView)
        addSubview(titleLab)
    }
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    override func layoutSubviews() {
        super.layoutSubviews()
        
        let max_v: CGFloat = 6
        
        if contentInsetOfTop > self.frame.size.height / max_v {
            contentInsetOfTop = 0.0
        }
        if contentInsetOfBottom >  self.frame.size.height / max_v {
            contentInsetOfBottom = 0.0
        }
        if spacing > self.frame.size.height / max_v {
            spacing = 0.0
        }
        
        let lab_height = titleLab.text!.calculateStringHeight(width: self.frame.size.width, font: titleLab.font)
        let img_height: CGFloat = self.frame.size.height - contentInsetOfTop - contentInsetOfBottom - lab_height - spacing
        
        let iv_y: CGFloat = contentInsetOfTop
        let iv_height: CGFloat = img_height
        let iv_width: CGFloat = iv_height
        let iv_x: CGFloat = 0.5 * (self.frame.size.width - iv_width)
        imgView.frame = CGRect.init(x: iv_x, y: iv_y, width: iv_width, height: iv_height)

        let tl_x: CGFloat = 0
        let tl_y: CGFloat = imgView.frame.maxY + spacing
        let tl_width: CGFloat = self.frame.size.width
        let tl_height: CGFloat = lab_height
        titleLab.frame = CGRect.init(x: tl_x, y: tl_y, width: tl_width, height: tl_height)
    }
    
    lazy var imgView: UIImageView = {
        let tempImgView = UIImageView()
        return tempImgView;
    }()
    lazy var titleLab: UILabel = {
        let tempTitleLab = UILabel()
        tempTitleLab.textAlignment = NSTextAlignment.center
        return tempTitleLab
    }()
}


extension String {
    ///
    /// Calculate the height of a string based on width and font size
    ///
    /// - parameter width: CGFloat
    /// - parameter font: UIFont
    ///
    /// - returns: The height of the calculated string
    ///
    fileprivate func calculateStringHeight(width: CGFloat, font: UIFont) -> CGFloat {
        let attrs = [NSAttributedString.Key.font: font]
        let tempRect = (self as NSString).boundingRect(with: CGSize(width: width, height: 0), options: NSStringDrawingOptions.usesLineFragmentOrigin, attributes: attrs, context: nil)
        return tempRect.size.height
    }
}
